# Operators - Data Manipulation Tools

## Basic Operators

- An operator is a symbol of the programming language which is able to operate on the values.
- Most widely recognizable basic operators (arithmetic) operations
        Addition
        Subtraction
        Multiplication
        Division
        Floor Division
        Remainder (Modulo)
        Exponentiation
    

##### Addition Operator ( + )
- The additional operator in Python is completely in line with mathematical standards


In [1]:
print(4+8)

12


##### Subtraction Operator ( - )

- Subtraction operator can be used in two ways - 
        1. For mathematical operations (binary operator)
        2. It can change the sign of a number. (unary operator)


In [6]:
print("As a binary operator: ", 45-40)

print("As a Unary operator: ", -4)

As a binary operator:  5
As a Unary operator:  -4


##### Multiplication Operator ( * )

- Multiplication Operator works completely in line with the mathematical standards

In [7]:
print(10*5)

50


##### Division Operator ( / )

- The value before the slash is a Dividend
- The value after the slash is a divisor
- The result produced by the division operator is always a float

###### How not to divide

- Don't perform any division by zero


In [9]:
print(6/3)

2.0


##### Floor Division or Integer Division ( // )

- Double slash is an integer division
- It returns the integer as a result (The results are always rounded )
- The rounding oes towards the lesser integer value.(1.6 will be rounded to 1)

In [10]:
print(6//3)

2


In [11]:
print(6//4)

1


##### Remainder / Modulo ( % )


- Modulo gives the remainder left after the integer division.


In [12]:
print(14%4)

2


##### Exponentiation ( ** )

- Exponentiation operator is used to raise the power of a numerical to certain power of our choice
- If both the arguments are integer, then the result will be integer.
- If both the arguments are float, then the result will be float. 

In [13]:
print(2**3)

8


## Operators and their priorities


Priority 1 : **

Priority 2 : +, - : Unary

Priority 3 : *, /, //, %

Priority 4 : +, - : Binary 

- Subexpressions in parantheses are always calculated first.
- Most Python operators have left-side binding, i.e., the calculation of the expression is conducted from left to right.
- The exponentiation operator uses right-sided binding.

In [14]:
print(2*3%5)

1


In [15]:
print((5 * ((25 % 13) + 100) / (2 * 13)) // 2)
    

10.0


## Summary

1. An expression is a combination of values (or variables, operators, calls to functions ‒ you will learn about them soon) which evaluates to a certain value, e.g., 1 + 2.

2. Operators are special symbols or keywords which are able to operate on the values and perform (mathematical) operations, e.g., the * operator multiplies two values: x * y.

3. Arithmetic operators in Python: + (addition), - (subtraction), * (multiplication), / (classic division ‒ always returns a float), % (modulus ‒ divides left operand by right operand and returns the remainder of the operation, e.g., 5 % 2 = 1), ** (exponentiation ‒ left operand raised to the power of right operand, e.g., 2 ** 3 = 2 * 2 * 2 = 8), // (floor/integer division ‒ returns a number resulting from division, but rounded down to the nearest whole number, e.g., 3 // 2.0 = 1.0)

4. A unary operator is an operator with only one operand, e.g., -1, or +3.

5. A binary operator is an operator with two operands, e.g., 4 + 5, or 12 % 5.

6. Some operators act before others - the hierarchy of priorities:

>> the ** operator (exponentiation) has the highest priority;
then the unary + and - (note: a unary operator to the right of the exponentiation operator binds more strongly, for example 4 ** -1 equals 0.25)
then: *, /, and %,
and finally, the lowest priority: binary + and -.

7. Subexpressions in parentheses are always calculated first, e.g., 15 - 1 * (5 * (1 + 2)) = 0.

8. The exponentiation operator uses right-sided binding, e.g., 2 ** 2 ** 3 = 256.



                        --______________--______________--______________--______________--